core/src/ncn_reward_router.rs

```diff
@@ -0,0 +1,574 @@
+            self.route_to_operator_rewards(operator_rewards)?;
+        }
+
+        // Vault Rewards
+        {
+            let operator_stake_weight = operator_snapshot.stake_weights();
+            let rewards_to_process: u64 = self.reward_pool();
+
+            for vault_operator_delegation in operator_snapshot.vault_operator_stake_weight().iter()
+            {
+                let vault = vault_operator_delegation.vault();
+                let vault_ncn_fee_group = vault_operator_delegation.ncn_fee_group();
+
+                let vault_reward_stake_weight = vault_operator_delegation
+                    .stake_weights()
+                    .ncn_fee_group_stake_weight(vault_ncn_fee_group)?;
+
+                let operator_reward_stake_weight =
+                    operator_stake_weight.ncn_fee_group_stake_weight(vault_ncn_fee_group)?;
+
+                let vault_reward = Self::calculate_vault_reward(
+                    vault_reward_stake_weight,
+                    operator_reward_stake_weight,
+                    rewards_to_process,
+                )?;
+
+                self.route_from_reward_pool(vault_reward)?;
+                self.route_to_vault_reward_route(vault, vault_reward)?;
+            }
+        }
+
+        // Operator gets any remainder
+        {
+            let leftover_rewards = self.reward_pool();
+
+            self.route_from_reward_pool(leftover_rewards)?;
+            self.route_to_operator_rewards(leftover_rewards)?;
+        }
+
+        Ok(())
+    }
+
+    // ------------------------ CALCULATIONS ------------------------
+    fn calculate_operator_reward(
+        fee_bps: u64,
+        rewards_to_process: u64,
+    ) -> Result<u64, TipRouterError> {
+        if fee_bps == 0 || rewards_to_process == 0 {
+            return Ok(0);
+        }
+
+        let precise_dao_fee_bps =
+            PreciseNumber::new(fee_bps as u128).ok_or(TipRouterError::NewPreciseNumberError)?;
+
+        let precise_max_bps =
+            PreciseNumber::new(MAX_BPS as u128).ok_or(TipRouterError::NewPreciseNumberError)?;
+
+        let precise_rewards_to_process = PreciseNumber::new(rewards_to_process as u128)
+            .ok_or(TipRouterError::NewPreciseNumberError)?;
+
+        let precise_dao_rewards = precise_rewards_to_process
+            .checked_mul(&precise_dao_fee_bps)
+            .and_then(|x| x.checked_div(&precise_max_bps))
+            .ok_or(TipRouterError::ArithmeticOverflow)?;
+
+        let floored_precise_dao_rewards = precise_dao_rewards
+            .floor()
+            .ok_or(TipRouterError::ArithmeticFloorError)?;
+
+        let dao_rewards_u128: u128 = floored_precise_dao_rewards
+            .to_imprecise()
+            .ok_or(TipRouterError::CastToImpreciseNumberError)?;
+        let dao_rewards: u64 = dao_rewards_u128
+            .try_into()
+            .map_err(|_| TipRouterError::CastToU64Error)?;
+
+        Ok(dao_rewards)
+    }
+
+    fn calculate_vault_reward(
```

Comment:
function names for this and calculate_operator_rewards seem off based on what they're returning

Reply:
updated names